File: memories/2024/exercise/java/day13/src/test/java/santamarket/model/SantamarketTest.java
package santamarket.model;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;

class SantamarketTest {

    @Test
    void noDiscount() {
        SantamarketCatalog catalog = new FakeCatalog();
        Product teddyBear = new Product("teddyBear", ProductUnit.EACH);
        catalog.addProduct(teddyBear, 1.0);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        sleigh.addItemQuantity(teddyBear, 3);

        ChristmasElf elf = new ChristmasElf(catalog);
        Receipt receipt = elf.checksOutArticlesFrom(sleigh);

        assertEquals(3.0, receipt.getTotalPrice(), 0.01);
    }

    @Test
    void tenPercentDiscount() {
        SantamarketCatalog catalog = new FakeCatalog();
        Product turkey = new Product("turkey", ProductUnit.KILO);
        catalog.addProduct(turkey, 2.0);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        sleigh.addItemQuantity(turkey, 2.0);

        ChristmasElf elf = new ChristmasElf(catalog);
        elf.addSpecialOffer(SpecialOfferType.TEN_PERCENT_DISCOUNT, turkey, 10.0);
        Receipt receipt = elf.checksOutArticlesFrom(sleigh);

        assertEquals(3.6, receipt.getTotalPrice(), 0.01);  // 4.0 - 0.4
    }

    @Test
    void threeForTwoDiscount() {
        SantamarketCatalog catalog = new FakeCatalog();
        Product teddyBear = new Product("teddyBear", ProductUnit.EACH);
        catalog.addProduct(teddyBear, 1.0);

        ChristmasElf elf = new ChristmasElf(catalog);
        elf.addSpecialOffer(SpecialOfferType.THREE_FOR_TWO, teddyBear, 0.0);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        sleigh.addItemQuantity(teddyBear, 3);

        Receipt receipt = elf.checksOutArticlesFrom(sleigh);

        assertEquals(2.0, receipt.getTotalPrice(), 0.01);  // Buy 3, pay for 2
    }
}

File: memories/2024/exercise/kotlin/day13/src/test/kotlin/santamarket/model/SantamarketTest.kt
package santamarket.model

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals

class SantamarketTest {

    @Test
    fun `no discount`() {
        val catalog = FakeCatalog()
        val teddyBear = Product("teddyBear", ProductUnit.EACH)
        catalog.addProduct(teddyBear, 1.0)

        val sleigh = ShoppingSleigh()
        sleigh.addItemQuantity(teddyBear, 3.0)

        val elf = ChristmasElf(catalog)
        val receipt = elf.checksOutArticlesFrom(sleigh)

        assertEquals(3.0, receipt.getTotalPrice(), 0.01)
    }

    @Test
    fun `ten percent discount`() {
        val catalog = FakeCatalog()
        val turkey = Product("turkey", ProductUnit.KILO)
        catalog.addProduct(turkey, 2.0)

        val sleigh = ShoppingSleigh()
        sleigh.addItemQuantity(turkey, 2.0)

        val elf = ChristmasElf(catalog)
        elf.addSpecialOffer(SpecialOfferType.TEN_PERCENT_DISCOUNT, turkey, 10.0)
        val receipt = elf.checksOutArticlesFrom(sleigh)

        assertEquals(3.6, receipt.getTotalPrice(), 0.01) // 4.0 - 0.4
    }

    @Test
    fun `three for two discount`() {
        val catalog = FakeCatalog()
        val teddyBear = Product("teddyBear", ProductUnit.EACH)
        catalog.addProduct(teddyBear, 1.0)

        val elf = ChristmasElf(catalog)
        elf.addSpecialOffer(SpecialOfferType.THREE_FOR_TWO, teddyBear, 0.0)

        val sleigh = ShoppingSleigh()
        sleigh.addItemQuantity(teddyBear, 3.0)

        val receipt = elf.checksOutArticlesFrom(sleigh)

        assertEquals(2.0, receipt.getTotalPrice(), 0.01) // Buy 3, pay for 2
    }
}
